queue/server.go

```diff
@@ -0,0 +1,121 @@
+package queue
+
+import (
+	"sync"
+
+	"context"
+	"errors"
+	"time"
+
+	"github.com/centrifuge/gocelery"
+	logging "github.com/ipfs/go-log"
+)
+
+const (
+	BlockHeightParam string = "BlockHeight"
+	TimeoutParam     string = "Timeout"
+)
+
+var log = logging.Logger("queue-server")
+
+// Config is an interface for queue specific configurations
+type Config interface {
+
+	// GetNumWorkers gets the number of background workers to initiate
+	GetNumWorkers() int
+
+	// GetWorkerWaitTime gets the worker wait time for a task to be available while polling
+	// increasing this may slow down task execution while reducing it may consume a lot of CPU cycles
+	GetWorkerWaitTimeMS() int
+}
+
+// TaskType is a task to be queued in the centrifuge node to be completed asynchronously
+type TaskType interface {
+
+	// TaskTypeName of the task
+	TaskTypeName() string
+}
+
+// TaskResult represents a result from a queued task execution
+type TaskResult interface {
+
+	// Get the result within a timeout from the queue task execution
+	Get(timeout time.Duration) (interface{}, error)
+}
+
+// Server represents the queue server currently implemented based on gocelery
+type Server struct {
+	config    Config
+	lock 	  sync.RWMutex
+	queue     *gocelery.CeleryClient
+	taskTypes []TaskType
+}
+
+// TaskTypeName of the queue server
+func (qs *Server) Name() string {
+	return "QueueServer"
+}
+
+// Start the queue server
+func (qs *Server) Start(ctx context.Context, wg *sync.WaitGroup, startupErr chan<- error) {
+	defer wg.Done()
+	qs.lock.Lock()
+	var err error
+	qs.queue, err = gocelery.NewCeleryClient(
+		gocelery.NewInMemoryBroker(),
+		gocelery.NewInMemoryBackend(),
+		qs.config.GetNumWorkers(),
+		qs.config.GetWorkerWaitTimeMS(),
+	)
+	if err != nil {
+		startupErr <- err
+	}
+	for _, task := range qs.taskTypes {
+		qs.queue.Register(task.TaskTypeName(), task)
+	}
+	// start the workers
+	qs.queue.StartWorker()
+	qs.lock.Unlock()
+
+	<-ctx.Done()
+	log.Info("Shutting down Queue server with context done")
+	qs.lock.Lock()
+	qs.queue.StopWorker()
+	qs.lock.Unlock()
+	log.Info("Queue server stopped")
+}
+
+// RegisterTaskType registers a task type on the queue server
+func (qs *Server) RegisterTaskType(name string, task interface{}) {
+	qs.taskTypes = append(qs.taskTypes, task.(TaskType))
```

Comment:
Lock is needed here since you are changing the state of qs.taskTypes.

Reply:
no need as its not being used across threads because node init is single threaded. Besides race detector is enabled on our CI it would detect if there is an issue.

Reply:
> no need as its not being used across threads because node init is single threaded. 
We cannot assume this will the be the case at a later time :)